fn.go

```diff
@@ -0,0 +1,230 @@
+package secp256k1
+
+/*
+
+// NOTE: Requires the gmp c library.
+#define USE_NUM_GMP
+
+#define HAVE___INT128
+#define USE_SCALAR_4X64
+#define USE_SCALAR_INV_BUILTIN
+
+#include "secp256k1/include/secp256k1.h"
+#include "secp256k1/src/util.h"
+#include "secp256k1/src/num_gmp_impl.h"
+#include "secp256k1/src/scalar.h"
+#include "secp256k1/src/scalar_impl.h"
+#include "secp256k1/src/scalar_4x64_impl.h"
+
+*/
+import "C"
+import (
+	"crypto/rand"
+	"fmt"
+	"io"
+	"math/big"
+	"unsafe"
+
+	"github.com/renproject/surge"
+)
+
+// Fn represents an element of the field defined by the prime N, where N is the
+// order of the elliptic curve group secp256k1.
+type Fn struct {
+	inner C.secp256k1_scalar
+}
+
+// NewFnFromUint returns a new field element equal to the given unsigned
+// integer.
+func NewFnFromUint(v uint16) Fn {
+	x := Fn{}
+	x.SetUint(v)
+	return x
+}
+
+// RandomFn returns a random field element.
+//
+// Panics: This function will panic if there was an error reading bytes from
+// the random source.
+func RandomFn() Fn {
+	x, err := RandomFnSafe()
+	if err != nil {
+		panic(fmt.Sprintf("could not generate random bytes: %v", err))
+	}
+	return x
+}
+
+// RandomFnSafe returns a random field element or an error.
+func RandomFnSafe() (Fn, error) {
+	var bs [32]byte
+	_, err := rand.Read(bs[:])
+	if err != nil {
+		return Fn{}, err
+	}
+	x := Fn{}
+
+	// This will reduce the value modulo N, so it does not matter if the bytes
+	// represent a number greater than N.
+	x.SetB32(bs[:])
+	return x, nil
+}
+
+// Clear sets the underlying data of the structure to zero. This will leave it
+// in a state which is a representation of the zero element.
+func (x *Fn) Clear() {
+	C.secp256k1_scalar_clear(&x.inner)
+}
+
+// Int returns a big.Int representation of the field element.
+func (x *Fn) Int() *big.Int {
+	ret := new(big.Int)
+	x.GetInt(ret)
+	return ret
+}
+
+// GetInt sets the given big.Int to be equal to the field element.
+func (x *Fn) GetInt(dst *big.Int) {
+	var bs [32]byte
+	x.GetB32(bs[:])
+	dst.SetBytes(bs[:])
+}
+
+// SetB32 sets the field element to be equal to the given byte slice,
+// interepreted as big endian. The field element will be reduced modulo N. The
+// argument `overflow` will be non zero if the bytes represented a number
```

Comment:
This comment will be confusing for people because `overflow` will not be visible to doc readers (it is a local variable, not an argument). Consider moving this comment into the function body.